### ISSUE TITLE
zos push succeeds on Kovan testnet and fails on mainnet, failed with error: Failed to check for transaction receipt: {}

### ISSUE THREAD

**Issue by klivin** [zeppelinos/zos#897]
npx zos --version
2.2.3

Trying to do a small upgrade to logic of one of our contracts.  Everything is successful using the zos push to Kovan.

Things fail on mainnet and we've spent over an ETH retrying to no avail.  Here's what happens:

npx zos bump 0.2.2
npx zos push
```
Compiling contracts with Truffle...

Compiling your contracts...
===========================
> Compiling ./contracts/IXyRequester.sol
> Compiling ./contracts/Migrations.sol
> Compiling ./contracts/PLCR/PLCRVoting.sol
> Compiling ./contracts/PLCR/attrstore/AttributeStore.sol
> Compiling ./contracts/PLCR/dll/DLL.sol
> Compiling ./contracts/XyBlockProducer.sol
> Compiling ./contracts/XyGovernance.sol
> Compiling ./contracts/XyParameterizer.sol
> Compiling ./contracts/XyPayOnDelivery.sol
> Compiling ./contracts/access/GovernorRole.sol
> Compiling ./contracts/access/Roles.sol
> Compiling ./contracts/introspection/ERC165.sol
> Compiling ./contracts/introspection/ERC165Checker.sol
> Compiling ./contracts/introspection/IERC165.sol
> Compiling ./contracts/staking/XyBond.sol
> Compiling ./contracts/staking/XyStakingConsensus.sol
> Compiling ./contracts/staking/XyStakingModel.sol
> Compiling ./contracts/tests/XyBlockProducerMock.sol
> Compiling ./contracts/tests/XyConsensusMock.sol
> Compiling ./contracts/tests/XyParameterizerMock.sol
> Compiling ./contracts/tests/XyPayOnDeliveryMock.sol
> Compiling ./contracts/tests/XyStakingMock.sol
> Compiling ./contracts/token/ERC20/IApprovalRecipient.sol
> Compiling ./contracts/token/ERC20/IERC20.sol
> Compiling ./contracts/token/ERC20/IXYOERC20.sol
> Compiling ./contracts/token/ERC20/SafeERC20.sol
> Compiling ./contracts/token/ERC20/XyERC20Token.sol
> Compiling ./contracts/token/ERC20/XyFaucet.sol
> Compiling ./contracts/utils/Address.sol
> Compiling ./contracts/utils/Initializable.sol
> Compiling ./contracts/utils/SafeMath.sol

    > compilation warnings encountered:

/Users/kevin/GitHub/xyo/dapp-scsc-solidity/contracts/staking/XyStakingModel.sol:423:5: Warning: Function state mutability can be restricted to pure
    function isBondedStake(bytes32) internal view returns (bool) {
    ^ (Relevant source part starts here and spans across multiple lines).

> Artifacts written to /Users/kevin/GitHub/xyo/dapp-scsc-solidity/build/contracts
> Compiled successfully using:
   - solc: 0.5.4+commit.9549d8ff.Emscripten.clang


Using session with network mainnet, sender address 0x6792B02f88b32C4FE8e31cfA41ae5aF44865f930, timeout 600 seconds
Validating contract XyGovernance
- Variables challenges (XyParameterizer), proposals (XyParameterizer), resolutions (XyGovernance), actions (XyGovernance) contain a struct or enum. These are not automatically checked for storage compatibility in the current version. See https://docs.zeppelinos.org/docs/writing_contracts.html#modifying-your-contracts for more info.
Validating contract XyBlockProducer
- Variable governors (GovernorRole) contains a struct or enum. These are not automatically checked for storage compatibility in the current version. See https://docs.zeppelinos.org/docs/writing_contracts.html#modifying-your-contracts for more info.
Validating contract XyStakingConsensus
- Variables stakeData (XyStakingModel), stakeeStake (XyStakingModel), stakerStake (XyStakingModel), requestsById (XyStakingConsensus), blocks (XyStakingConsensus) contain a struct or enum. These are not automatically checked for storage compatibility in the current version. See https://docs.zeppelinos.org/docs/writing_contracts.html#modifying-your-contracts for more info.
Validating contract PLCRVoting
- Variables pollMap (PLCRVoting), dllMap (PLCRVoting), store (PLCRVoting) contain a struct or enum. These are not automatically checked for storage compatibility in the current version. See https://docs.zeppelinos.org/docs/writing_contracts.html#modifying-your-contracts for more info.
Validating contract XyBond
- Variables governors (GovernorRole), bond (XyBond) contain a struct or enum. These are not automatically checked for storage compatibility in the current version. See https://docs.zeppelinos.org/docs/writing_contracts.html#modifying-your-contracts for more info.
Current version 0.2.1
Creating new version 0.2.2
Uploading XyGovernance contract as XyGovernance
Setting implementation of XyGovernance in directory...
Uploading XyBlockProducer contract as XyBlockProducer
Setting implementation of XyBlockProducer in directory...
Uploading XyStakingConsensus contract as XyStakingConsensus
Setting implementation of XyStakingConsensus in directory...
Uploading PLCRVoting contract as PLCRVoting
Setting implementation of PLCRVoting in directory...
Uploading XyBond contract as XyBond
Setting implementation of XyBond in directory...
Setting XyGovernance implementation 0x8637af4d686404e7E9bbD96EeD53dF9E20110334...
Setting XyBlockProducer implementation 0x6c9dE9D7199a24999a0ccAcADF417850cB3DF4b8...
XyBlockProducer deployment failed with error: Failed to check for transaction receipt:
{}
XyBond deployment failed with error: Failed to check for transaction receipt:
{}
XyStakingConsensus deployment failed with error: Failed to check for transaction receipt:
{}
XyGovernance deployment failed with error: Failed to check for transaction receipt:
{}
PLCRVoting deployment failed with error: Failed to check for transaction receipt:
{}
Updated zos.mainnet.json

```

This pretty much wipes our zos.mainnet.json which now looks like:

```
{
  "contracts": {},
  "solidityLibs": {
    "DLL": {
      "address": "0xe9b3378C206c93D8811e04bEd749784CEdD23E4B",
      "constructorCode": "6105ff610030600b82828239805160001a6073146000811461002057610022565bfe5b5030600052607381538281f3fe",
      "bodyBytecodeHash": "0524c747135bb99ca95f6d5cb83a27797fe03cc72961fe3bc1f7b14665d47731",
      "localBytecodeHash": "7b85ab7075d24dd476ba4afb4318200b98e1edebc09e2e53169a6106097cefea",
      "deployedBytecodeHash": "7b85ab7075d24dd476ba4afb4318200b98e1edebc09e2e53169a6106097cefea"
    }
  },
  "proxies": {
    "dapp-scsc-solidity/XyBlockProducer": [
      {
        "address": "0xd3034c290E19959Fdc18E5597F375CA51BFB0c41",
        "version": "0.1.1",
        "implementation": "0x9b9b129dEc722CBafa38D4E60E7dB9Fd5A2DeF75"
      }
    ],
    "dapp-scsc-solidity/PLCRVoting": [
      {
        "address": "0x72bCDb36d1545FEA06905a1bb4998424580AAee6",
        "version": "0.1.1",
        "implementation": "0x7C1013EeF811d00cB090b9118Dd0742961c04D6F"
      }
    ],
    "dapp-scsc-solidity/XyGovernance": [
      {
        "address": "0x01925d0fFE4a6a6162B51ba611e3D4780Fc2dF42",
        "version": "0.1.1",
        "implementation": "0xeb735bE25A5bB8138A07ab3d4B4C998ee28f86f9"
      }
    ],
    "dapp-scsc-solidity/XyStakingConsensus": [
      {
        "address": "0x0242514106114DEaA99Fd81574142c36Edb03B6D",
        "version": "0.2.1",
        "implementation": "0xC150EBAFce052C216D609D56DfbEDf817F9DB339"
      }
    ],
    "dapp-scsc-solidity/XyBond": [
      {
        "address": "0x1a024A698EEBdB86ccf3fCaF2F589839bdc066AD",
        "version": "0.2.1",
        "implementation": "0x81391F9821660172aeF6DF728784deC5aBf46a42"
      }
    ]
  },
  "zosversion": "2.2",
  "version": "0.2.2",
  "dependencies": {
    "openzeppelin-eth": {
      "package": "0x778dddf23ec1b5cb18394c6c110480caadb3b0f6",
      "version": "2.1.3"
    }
  },
  "proxyAdmin": {
    "address": "0x851Bf4df35CfE742A3Cae42601bDD1fe68Df03dC"
  },
  "package": {
    "address": "0x6D9cC7FD835Ae31eA15aDd31918953e5C33d2797"
  },
  "app": {
    "address": "0xa87838152519c7571A64893C65E3441347D31Bf2"
  },
  "provider": {
    "address": "0x3e03a5AD2e98152BD8eAcdF08D9eF5bd2e77d689"
  },
  "frozen": false
}
```

**Comment by spalladino** [comment]
Hey Kevin, could you share some info regarding the node you are using for connecting to mainnet? Is it a private node, or an infura public node? If private, is it geth or parity, and which version? Also, could you run a `getBlock('latest')` and check against etherscan that the block number matches the latest one, and the node is not our of sync?

**Comment by klivin** [comment]
I am connecting via Infura, here's my truffle config:
```
 mainnet: {
      network_id: `1`,
      provider: () => {
        const wallet = new HDWalletProvider(mnemonic, `https://mainnet.infura.io/${infuraKey}`)
        const nonceTracker = new NonceTrackerSubprovider()
        wallet.engine._providers.unshift(nonceTracker)
        nonceTracker.setEngine(wallet.engine)
        return wallet
      },
      gas: 6986331,
      gasPrice: 25000000000
    }
```

I have tried with and without the `NonceTrackerSubprovider`

**Comment by spalladino** [comment]
Apparently this bug has been reported [several times on web3js](https://github.com/ethereum/web3.js/issues?utf8=%E2%9C%93&q=Failed+to+check+for+transaction+receipt), and the comments show several different (and never consistent) solutions. All in all, it seems like a problem in web3 when trying to retrieve the receipt from the node.

I'll try to reproduce it locally during this week, but in the meantime, can I ask you to test with a local node (with fast sync), instead of Infura? The fact that it only happens on mainnet may be pointing to an issue with a cluttered network, and running from a local node might help.

**Comment by klivin** [comment]
Thanks for the info @spalladino.  I noticed that I used Infura endpoint without a `v3` in it.  I wonder if that could help with the problem, we also updated truffle to 5.0.18.  I will wait for a low traffic time tonight to try again.  Also, is it best to revert to the previous working zos.mainnet.json before doing the next npx zos push?  I recall there being an automatic dry-run in previous versions, I'll look into that as well.  Here's our package.json for info:

```
{
  "name": "@xyo-network/dapp-scsc-solidity",
  "version": "1.1.3",
  "description": "XYO Staked Consensus Smart Contract",
  "directories": {
    "test": "test"
  },
  "scripts": {
    "build": "dapploy",
    "deploy": "dapploy -p",
    "chain": "ganache-cli"
  },
  "author": "Kevin Weiler",
  "license": "MIT",
  "devDependencies": {
    "babel-eslint": "^10.0.1",
    "babel-polyfill": "^6.26.0",
    "babel-preset-es2015": "^6.24.1",
    "babel-preset-stage-2": "^6.24.1",
    "babel-preset-stage-3": "^6.24.1",
    "babel-register": "^6.26.0",
    "bignumber.js": "^8.1.1",
    "chai": "^4.1.2",
    "chai-as-promised": "^7.1.1",
    "chai-bignumber": "^3.0.0",
    "dotenv": "^8.0.0",
    "ethereumjs-abi": "^0.6.6",
    "ganache-cli": "^6.4.3",
    "openzeppelin-solidity": "^2.1.3",
    "openzeppelin-test-helpers": "^0.3.2",
    "solc": "^0.5.8",
    "solium": "^1.2.4",
    "truffle": "5.0.18",
    "truffle-hdwallet-provider": "^1.0.9"
  },
  "dependencies": {
    "bs58": "^4.0.1",
    "ethereumjs-util": "^6.1.0",
    "openzeppelin-eth": "^2.1.3",
    "web3-provider-engine": "^15.0.0",
    "zos": "^2.2.3",
    "zos-lib": "^2.2.3"
  },
  "files": [
    "build",
    "contracts",
    "test",
    "zos.json",
    "zos.*.json"
  ],
  "snyk": true
}
```

**Comment by spalladino** [comment]
Hey @klivin, I've just tried to reproduce the issue, hitting the Infura mainnet endpoint v3 with a 20GWei gas price, and the deployment of a published project with 10 contracts went fine. FWIW, I'm using the following versions of dependencies:

```
"truffle": "5.0.7",
"truffle-hdwallet-provider": "1.0.4",
```

With the following truffle config:
```
  "mainnet-infura": {
      provider: () => new HDWalletProvider(mnemonicMainnet, "https://mainnet.infura.io/v3/APIKEY"),
      network_id: 3,
      gasPrice: 20e9
    }
```

Let me know if it works on your end after changing the infura endpoint!

**Comment by klivin** [comment]
Hi @spalladino, I notice you are using network_id: 3 in your config, which is Ropsten, did you mean 1?  The deployment works perfectly on testnet, Kovan, for me.

As for mainnet, I tried with the new mainnet infura `v3` endpoint and zos push got a bit further but ended with a new error (full printout at bottom):
`XyStakingConsensus deployment failed with error: The contract code couldn't be stored, please check your gas limit.`

We are deploying 5 contracts and I find it odd that my etherscan report shows the 5 contracts created, and there's 4 followup transactions, apparently it ran out of gas on the last one?
https://etherscan.io/address/0x6792b02f88b32c4fe8e31cfa41ae5af44865f930

I wish it would keep track and start from where it left off in the transactions, as this is expensive...  I am hopeful though and it was so close...below is the printout of the latest attempt.  I am also busy syncing a local node (it's been 2 days and I'm half way) to see if that will help at all.  Otherwise I'm open for suggestions.  I used truffle 5.0.18, and I could try on the 5.0.7.

```
Current version 0.2.1
Creating new version 0.2.2
Uploading XyGovernance contract as XyGovernance
Setting implementation of XyGovernance in directory...
Uploading XyBlockProducer contract as XyBlockProducer
Setting implementation of XyBlockProducer in directory...
Uploading XyStakingConsensus contract as XyStakingConsensus
Setting implementation of XyStakingConsensus in directory...
Uploading PLCRVoting contract as PLCRVoting
Setting implementation of PLCRVoting in directory...
Uploading XyBond contract as XyBond
Setting implementation of XyBond in directory...
Setting XyGovernance implementation 0xfDa25EC2dD8eec3421c2B2cDB4E4ED9A719C0fF5...
Setting XyBlockProducer implementation 0x71b762978197290Cf00716eC21fC4F373145666a...
Setting PLCRVoting implementation 0xfF6Aa96233Fd4a68B46d242cA7Cda34F101587F0...
Setting XyBond implementation 0x9D27aA32Acd0456fB4c40699f0E8592625169F3b...
Implementation set: 0xfDa25EC2dD8eec3421c2B2cDB4E4ED9A719C0fF5
Implementation set: 0xfDa25EC2dD8eec3421c2B2cDB4E4ED9A719C0fF5
Implementation set: 0x71b762978197290Cf00716eC21fC4F373145666a
Implementation set: 0x71b762978197290Cf00716eC21fC4F373145666a
Implementation set: 0xfF6Aa96233Fd4a68B46d242cA7Cda34F101587F0
Implementation set: 0xfF6Aa96233Fd4a68B46d242cA7Cda34F101587F0
Implementation set: 0x9D27aA32Acd0456fB4c40699f0E8592625169F3b
Implementation set: 0x9D27aA32Acd0456fB4c40699f0E8592625169F3b
XyStakingConsensus deployment failed with error: The contract code couldn't be stored, please check your gas limit.
Updated zos.mainnet.json
```

**Comment by klivin** [comment]
Well, it turns out it does start where it left off, and it ultimately succeeded. The only thing I changed was downgrading to:
`
"truffle": "5.0.7",
"truffle-hdwallet-provider": "1.0.4",
`
 I also changed gasPrice to
`gasPrice: 20e9`

🤷‍♂

**Comment by spalladino** [comment]
Glad to hear it worked, @klivin! We'll continue reviewing this to see if we can reproduce these scenarios, which are sometimes caused by unreliable infrastructure or dependencies.

>  I notice you are using network_id: 3 in your config, which is Ropsten, did you mean 1? The deployment works perfectly on testnet, Kovan, for me.
Actually, that was a copy-paste error when I set up the infura connection, but it seems the hdwallet provider ignores that parameter and connects anyway to whichever network it is. All txs went through to mainnet (I checked them on etherscan).
